File: LinkedListCycleII.java
/**
 * LeetCode 142 Linked List Cycle II
 */
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {
        if (head == null || head.next == null) {
            return null;
        }
        Set<ListNode> container = new HashSet<>();
        while (head != null) {
            if (container.contains(head)) {
                return head;
            } else {
                container.add(head);
                head = head.next;
            }
        }
        return null;
    }
}

File: LinkedListCycleII.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func detectCycle(head *ListNode) *ListNode {
	slow ,fast := head,head
	for (fast != nil) {
		if (fast.Next == nil) {
			return nil
		}
		slow = slow.Next
		fast = fast.Next.Next

		if (fast == slow) {
			ptr := head
			for (ptr != slow) {
				ptr = ptr.Next
				slow = slow.Next
			}
			return ptr
		}
	}
	return nil
}
